docs/spikes/2020-04-how-to-migrate-lb3-components.md

```diff
@@ -0,0 +1,626 @@
+# How to migrate LB3 components
+
+_GitHub issue: https://github.com/strongloop/loopback-next/issues/4099_
+
+## Out of scope
+
+Migration of the following component is out of scope of this spike:
+
+- [OAuth2](https://loopback.io/doc/en/lb3/OAuth-2.0.html) - will be covered by
+  auth&auth migration guide
+- [Passport](https://loopback.io/doc/en/lb3/Third-party-login-using-Passport.html) -
+  will be covered by auth&auth migration guide
+- [Synchronization](https://loopback.io/doc/en/lb3/Synchronization.html) - won't
+  be implemented in LB4. It's important to mention this fact in the migration
+  guide.
+
+## Table of contents
+
+- [Techniques](#techniques)
+  - [General](#general)
+  - [Models, Entities and Repositories](#models-entities-and-repositories)
+  - [REST API](#rest-api)
+  - [Services (local and remote)](#services-local-and-remote)
+  - [API transports](#api-transports)
+  - [Authentication & authorization](#authentication--authorization)
+  - [Introspection](#introspection)
+- [Migration of component layout & instructions from mounting to an app](#migration-of-component-layout--instructions-from-mounting-to-an-app)
+  - [LB3 component layout & mounting](#lb3-component-layout--mounting)
+  - [Migration to LB4](#migration-to-lb4)
+- [Overview of existing LB3 components](#overview-of-existing-lb3-components)
+  - [Push notifications](#push-notifications)
+  - [Storage component](#storage-component)
+  - [Logging components](#logging-components)
+  - [Access control](#access-control)
+  - [Connectors](#connectors)
+  - [Custom transports](#custom-transports)
+  - [Inspecting models](#inspecting-models)
+  - [Unsorted](#unsorted)
+  - [Mixins](#mixins)
+
+## Techniques
+
+### General
+
+1. How to structure the extension code to receive user-provided configuration &
+   target app instance. What are the instructions for adding a LB4 component to
+   a LB4 app.
+
+2. A context shared by all parts of the application, allowing different layers
+   to store and retrieve values like "the current user". In LB3, we have
+   `loopback-context` based on continuation-local-storage and `options`-based
+   context propagation.
+
+3. How to migrate a component exporting mixins. See
+   [Migrating model mixins ](https://loopback.io/doc/en/lb4/migration-models-mixins.html)
+   for migration guide for app developers (mixin consumers).
```

Comment:
Maybe describe what a LB3 component can contribute to the application when it's mounted?

Reply:
Maybe a digram would help if it shows the handshake between an application and a component as well as typical artifacts exported from a component.

Reply:
To some extent, a LB3 application asks its components to extend/patch the app. In contrast, a LB4 application imports bindings (representing the component's contribution to the app) from its components.

Reply:
Those are all good suggestions for the migration guide, I'll capture them in a note in this document. Please note that I am not that far yet, here I am just collecting and describinh different techniques used by the components.